test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc

```diff
@@ -0,0 +1,187 @@
+            5678);
+}
+
+TEST(FakeStepController, IsConvergedInitiallyReturnsFalse) {
+  FakeStepController step_controller(FakeStepControllerConfig{}, CommandLineOptions{});
+  EXPECT_FALSE(step_controller.IsConverged());
+}
+
+TEST(FakeStepController, IsConvergedReturnsFalseAfterBenchmarkResultWithoutPositiveScore) {
+  FakeStepController step_controller(FakeStepControllerConfig{}, CommandLineOptions{});
+  BenchmarkResult benchmark_result;
+  step_controller.UpdateAndRecompute(benchmark_result);
+  EXPECT_FALSE(step_controller.IsConverged());
+}
+
+TEST(FakeStepController, IsConvergedReturnsTrueAfterBenchmarkResultWithPositiveScore) {
+  FakeStepController step_controller(FakeStepControllerConfig{}, CommandLineOptions{});
+  BenchmarkResult benchmark_result;
+  MetricEvaluation* evaluation = benchmark_result.mutable_metric_evaluations()->Add();
+  evaluation->set_threshold_score(1.0);
+  step_controller.UpdateAndRecompute(benchmark_result);
+  EXPECT_TRUE(step_controller.IsConverged());
+}
+
+TEST(FakeStepController, IsDoomedReturnsFalseAfterSuccessfulBenchmarkResult) {
+  FakeStepController step_controller(FakeStepControllerConfig{}, CommandLineOptions{});
+  BenchmarkResult benchmark_result;
+  benchmark_result.mutable_status()->set_code(::grpc::OK);
+  step_controller.UpdateAndRecompute(benchmark_result);
+  std::string doomed_reason;
+  EXPECT_FALSE(step_controller.IsDoomed(doomed_reason));
+}
+
+TEST(FakeStepController, IsDoomedDoesNotWriteDoomedReasonAfterSuccessfulBenchmarkResult) {
+  FakeStepController step_controller(FakeStepControllerConfig{}, CommandLineOptions{});
+  BenchmarkResult benchmark_result;
+  benchmark_result.mutable_status()->set_code(::grpc::OK);
+  step_controller.UpdateAndRecompute(benchmark_result);
+  std::string doomed_reason = "untouched";
+  ASSERT_FALSE(step_controller.IsDoomed(doomed_reason));
+  EXPECT_EQ(doomed_reason, "untouched");
+}
+
+TEST(FakeStepController, IsDoomedReturnsTrueAfterFailedBenchmarkResult) {
+  FakeStepController step_controller(FakeStepControllerConfig{}, CommandLineOptions{});
+  BenchmarkResult benchmark_result;
+  benchmark_result.mutable_status()->set_code(::grpc::INTERNAL);
+  benchmark_result.mutable_status()->set_message("error from nighthawk");
+  step_controller.UpdateAndRecompute(benchmark_result);
+  std::string doomed_reason;
+  EXPECT_TRUE(step_controller.IsDoomed(doomed_reason));
+}
+
+TEST(FakeStepController, IsDoomedSetsDoomedReasonToStatusMessageAfterFailedBenchmarkResult) {
+  FakeStepController step_controller(FakeStepControllerConfig{}, CommandLineOptions{});
+  BenchmarkResult benchmark_result;
+  benchmark_result.mutable_status()->set_code(::grpc::INTERNAL);
+  benchmark_result.mutable_status()->set_message("error from nighthawk");
+  step_controller.UpdateAndRecompute(benchmark_result);
+  std::string doomed_reason;
+  ASSERT_TRUE(step_controller.IsDoomed(doomed_reason));
+  EXPECT_EQ(doomed_reason, "error from nighthawk");
```

Comment:
"error from nighthawk" might be reduplicated by using a constant?

Reply:
Done